test/chat/db/scope/cargo_key_scope_test.exs

```diff
@@ -1,41 +1,129 @@
+    room_key = Identity.pub_key(room_identity)
+
+    generate_dialogs_and_cargo_room_invite([admin, john], operator, room_identity)
+    user_c = User.login("UserC") |> register_and_put_user()
+    generate_dialogs_and_cargo_room_invite([user_c], operator, room_identity)
+
+    user_d = User.login("UserD") |> register_and_put_user()
+    generate_dialogs_and_cargo_room_invite([user_d], user_c, room_identity)
+
+    assert_keys_for_cargo_keys(room_key, [checkpoint1_key, checkpoint2_key], 4)
+  end
+
+  defp setup_test_data do
+    {User.login("admin"), User.login("John"),
+     "Operator" |> User.login() |> register_and_put_user()}
+  end
 
+  defp assert_keys_for_cargo_keys(room_key, checkpoint_keys, expected_count) do
     Process.sleep(100)
+    keys = KeyScope.get_cargo_keys(Chat.Db.db(), room_key, checkpoint_keys)
+
+    expected_keys = %{
+      dialog_message: expected_count,
+      dialogs: expected_count,
+      room_invite: expected_count,
+      room_invite_index: expected_count,
+      rooms: 1
+    }
+
+    assert expected_keys == filter_and_count_keys(keys)
+  end
+
+  defp filter_and_count_keys(keys) do
+    keys
+    |> MapSet.to_list()
+    |> Enum.frequencies_by(&elem(&1, 0))
+    |> Map.drop([:users])
+  end
 
-    # Chat.Db.db()
-    # |> CubDB.select()
-    # |> Stream.map(fn {k, _} -> k end)
-    # |> Enum.to_list()
-    # |> Enum.frequencies_by(&elem(&1, 0))
-    # |> IO.inspect(label: "db", pretty: true)
+  defp generate_checkpoints(identities) do
+    data =
+      Enum.reduce(identities, %{keys: [], cards: []}, fn identity, acc ->
+        register_and_put_user(identity)
+        card = Chat.Card.from_identity(identity)
 
-    keys = KeyScope.get_cargo_keys(Chat.Db.db(), room_key, [bot_key])
+        %{
+          keys: [Identity.pub_key(identity) | acc.keys],
+          cards: [card | acc.cards]
+        }
+      end)
 
-    assert %{dialog_message: 1, dialogs: 1, room_invite: 1, room_invite_index: 1, rooms: 1} ==
-             keys
-             |> MapSet.to_list()
-             |> Enum.frequencies_by(&elem(&1, 0))
-             |> Map.drop([:users])
+    store_cargo_settings_checkpoints(data.cards)
+
+    data.keys
+  end
+
+  defp store_cargo_settings_checkpoints(cards) do
+    :ok =
+      AdminRoom.store_cargo_settings(%{
+        AdminRoom.get_cargo_settings()
+        | checkpoints: cards
+      })
+  end
+
+  defp register_and_put_user(identity) do
+    identity |> tap(&User.register/1) |> tap(&User.UsersBroker.put/1)
+  end
+
+  defp generate_cargo_room(user) do
+    {room_identity, _room} = Rooms.add(user, "Cargo room", :cargo)
+    room_identity
+  end
+
+  defp generate_dialogs_and_cargo_room_invite(checkpoints, user, room_identity) do
+    dialogs =
+      for checkpoint <- checkpoints,
+          into: [],
+          do: Dialogs.find_or_open(user, checkpoint |> Chat.Card.from_identity())
+
+    Enum.each(dialogs, fn dialog -> create_and_add_room_invite(room_identity, dialog, user) end)
```

Comment:
`for` and `Enum.each`  doing the same

Why to use both ?

Reply:
right, function from each moved to for